DirectProgramming/C++SYCL_FPGA/Tutorials/Features/IPAuthoring_interfaces_overview/host-pipes/src/vector_add.cpp

```diff
@@ -0,0 +1,105 @@
+#include <iostream>
+
+// oneAPI headers
+
+#include <sycl/ext/intel/experimental/pipes.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>  // sycl.hpp comes first to prevent compiler crash
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class SimpleVAdd_pipes;
+
+template <typename pipeA, typename pipeB, typename pipeC>
```

Comment:
Rather than templating the kernel on the host pipes, i think it would be more readable to declare the host pipe types above the kernel, then just refer to them directly, like this:

```c++
using InputPipeA = ...;
using InputPipeB = ...;
using InputPipeC = ...; 

struct SimpleVAddKernel_pipes {
  int len;

  void operator()() const {
    [[intel::initiation_interval(1)]]  //
    for (int idx = 0; idx < len; idx++) {
      int a_val = InputPipeA::read();
      int b_val = InputPipeB::read();
      int sum = a_val + b_val;
      OutputPipeC::write(sum);
    }
  }
};
...

 q.single_task<SimpleVAdd_pipes>(SimpleVAddKernel_pipes{count}).wait();